simpeg/directives/_regularization.py

```diff
@@ -0,0 +1,427 @@
+from __future__ import annotations
+
+import warnings
+
+import numpy as np
+from dataclasses import dataclass
+
+from .directives import InversionDirective, UpdatePreconditioner, BetaSchedule
+from simpeg.regularization import Sparse, BaseSparse, SmoothnessFirstOrder
+from simpeg.utils import validate_integer, validate_float
+
+
+@dataclass
+class IRLSMetrics:
+    """
+    Data class to store metrics used by the IRLS algorithm.
+
+    Parameters
+    ----------
+
+    input_norms: List of norms temporarily stored during the initialization.
+    irls_iteration_count: Number of IRLS iterations.
+    start_irls_iter: Iteration number when the IRLS process started.
+    f_old: Previous value of the regularization function.
+    """
+
+    input_norms: list[float] = (2.0, 2.0, 2.0, 2.0)
+    irls_iteration_count: int = 0
+    start_irls_iter: int | None = None
+    f_old: float = 0.0
+
+
+class UpdateIRLS(BetaSchedule):
+    """
+    Directive to control the IRLS iterations for regularization.Sparse.
+
+    Parameters
+    ----------
+    chifact_start: float
+        Starting chi factor for the IRLS iterations.
+    chifact_target: float
+        Target chi factor for the IRLS iterations.
+    irls_cooling_factor: float
+        Factor to cool the IRLS threshold epsilon.
+    f_min_change: float
+        Minimum change in the regularization function to continue the IRLS iterations.
+    max_irls_iterations: int
+        Maximum number of IRLS iterations.
+    misfit_tolerance: float
+        Tolerance for the target misfit.
+    percentile: int
+        Percentile of the function values used to determine the initial IRLS threshold.
+    verbose: bool
+        Print information to the screen.
+    """
+
+    def __init__(
+        self,
+        coolingRate: int = 1,
+        coolingFactor: float = 2.0,
```

Comment:
Moreover, since we are only inheriting `BetaSchedule` to use the `coolingFactor` and `coolingRate` properties, wouldn't be better to just implement those here (for `cooling_factor` and `cooling_rate`)? I know we would be repeating some code, but I think it worth for making the code cleaner: we are overriding the `endIter` method which is the core of the `BetaSchedule` directive, so the inheritance is only due to reuse of two properties.